src/tr1/game/output/sprites.c

```diff
@@ -0,0 +1,454 @@
+            current_vertex++;
+        }
+    }
+
+    glBindBuffer(GL_ARRAY_BUFFER, m_LevelData.shade_vbo);
+    GFX_TRACK_SUBDATA(
+        glBufferSubData, GL_ARRAY_BUFFER,
+        room_batch->quad_start * M_QUAD_VERTICES * sizeof(M_SPRITE_SHADE),
+        room_batch->quad_count * M_QUAD_VERTICES * sizeof(M_SPRITE_SHADE),
+        &m_LevelData.shade_vbo_data[room_batch->quad_start * M_QUAD_VERTICES]);
+}
+
+void M_GLDebug(
+    GLenum source, GLenum type, GLuint id, GLenum severity, GLsizei length,
+    const GLchar *message, const void *userParam)
+{
+    LOG_INFO("%d %s", source, message);
+}
+
+void Output_Sprites_Init(void)
+{
+    glDebugMessageCallback(M_GLDebug, nullptr);
+    glEnable(GL_DEBUG_OUTPUT_SYNCHRONOUS);
+
+    GFX_GL_Program_Init(&m_Program);
+    GFX_GL_Program_AttachShader(
+        &m_Program, GL_VERTEX_SHADER, "shaders/sprites.glsl",
+        GFX_Context_GetConfig()->backend);
+    GFX_GL_Program_AttachShader(
+        &m_Program, GL_FRAGMENT_SHADER, "shaders/sprites.glsl",
+        GFX_Context_GetConfig()->backend);
+    GFX_GL_Program_FragmentData(&m_Program, "outColor");
+    GFX_GL_Program_Link(&m_Program);
+
+    const char *const uniform_names[] = {
+        [M_UNIFORM_TEX_ATLAS] = "uTexture",
+        [M_UNIFORM_TEX_FRAMES] = "uFrame",
+        [M_UNIFORM_SMOOTHING_ENABLED] = "uSmoothingEnabled",
+        [M_UNIFORM_BRIGHTNESS_MULTIPLIER] = "uBrightnessMultiplier",
+        [M_UNIFORM_VIEWPORT_SIZE] = "uViewportSize",
+        [M_UNIFORM_PROJECTION_MATRIX] = "uMatProjection",
+        [M_UNIFORM_MODEL_MATRIX] = "uMatModelView",
+        [M_UNIFORM_WIBBLE_OFFSET] = "uWibbleOffset",
+        [M_UNIFORM_FOG] = "uFog",
+    };
+    for (int32_t i = 0; i < M_UNIFORM_NUMBER_OF; i++) {
+        m_Uniforms[i] =
+            GFX_GL_Program_UniformLocation(&m_Program, uniform_names[i]);
+        GFX_GL_CheckError();
+    }
+
+    GFX_GL_Program_Bind(&m_Program);
+    glUniform1i(m_Uniforms[M_UNIFORM_TEX_ATLAS], 0);
+    glUniform1i(m_Uniforms[M_UNIFORM_TEX_FRAMES], 1);
+}
+
+static void M_FreeBuffers(void)
+{
+    glBindVertexArray(0);
+    glBindBuffer(GL_ARRAY_BUFFER, 0);
+    if (m_LevelData.vao != 0) {
+        glDeleteVertexArrays(1, &m_LevelData.vao);
+        m_LevelData.vao = 0;
+    }
+    if (m_LevelData.geom_vbo != 0) {
+        glDeleteBuffers(1, &m_LevelData.geom_vbo);
+        m_LevelData.geom_vbo = 0;
+    }
+    if (m_LevelData.shade_vbo != 0) {
+        glDeleteBuffers(1, &m_LevelData.shade_vbo);
+        m_LevelData.shade_vbo = 0;
+    }
+    Memory_FreePointer(&m_LevelData.room_batches);
+    Memory_FreePointer(&m_LevelData.geom_vbo_data);
+    Memory_FreePointer(&m_LevelData.shade_vbo_data);
+}
+
+static void M_PrepareLevelBuffers(void)
+{
+    m_LevelData.total_vertex_count = 0;
+    m_LevelData.room_batch_count = Room_GetCount();
+    m_LevelData.texture_batch_count = 0;
+    for (int32_t i = 0; i < Room_GetCount(); i++) {
+        const ROOM *const room = Room_Get(i);
```

Comment:
Is this OK with flipmaps? E.g., when a room is flipped, the "other" room's mesh becomes the current room's and vice versa, but their indices remain the same. I think that means that if a normal room has sprites and a flipped room doesn't the sprites would remain visible in the flip map.